File: client/actionmenu.lua
--[[
     ____   ____ _____ _____   _   _____  ________      ________ _      ____  _____  __  __ ______ _   _ _______ 
    |  _ \ / __ \_   _|_   _| | | |  __ \|  ____\ \    / /  ____| |    / __ \|  __ \|  \/  |  ____| \ | |__   __|
    | |_) | |  | || |   | |   | | | |  | | |__   \ \  / /| |__  | |   | |  | | |__) | \  / | |__  |  \| |  | |   
    |  _ <| |  | || |   | |   | | | |  | |  __|   \ \/ / |  __| | |   | |  | |  ___/| |\/| |  __| | . ` |  | |   
    | |_) | |__| || |_ _| |_  | | | |__| | |____   \  /  | |____| |___| |__| | |    | |  | | |____| |\  |  | |   
    |____/ \____/_____|_____| | | |_____/|______|   \/   |______|______\____/|_|    |_|  |_|______|_| \_|  |_|   
                              | |                                                                                
                              |_|               UI ELEMENTS
]]

--- Client side action menu handling
-- @script client/action_menu.lua

--- @section Variables

--- Active flag
local active = false

--- @section Local functions

--- Process and filter menu options
-- @param options table: Received options to build the menu.
local function filter_menu(options)
    local processed_options = {}
    for _, option in ipairs(options) do
        local processed_option = deep_copy(option)
        if type(processed_option.can_interact) == 'function' then
            processed_option.interactable = processed_option.can_interact()
        else
            processed_option.interactable = true
        end
        if processed_option.action_type == 'function' and type(processed_option.action) == 'function' then
            register_function(processed_option.label, processed_option.action)
            processed_option.action = processed_option.label
        end
        if processed_option.submenu then
            processed_option.submenu = filter_menu(processed_option.submenu)
        end
        processed_options[#processed_options + 1] = processed_option
    end
    return processed_options
end

--- Opens the menu with given data.
-- @param menu_data table: The incoming menu data to open.
local function open_menu(menu_data)
    if active then return end
    active = true
    SetNuiFocus(true, true)
    SendNUIMessage({
        action = 'create_action_menu',
        menu = filter_menu(menu_data)
    })
end

--- @section NUI callbacks

--- Closes menu from UI actions.
-- @param data table: Received data from the UI.
RegisterNUICallback('close_action_menu', function(data)
    SetNuiFocus(false, false)
    active = false
end)

--- Event to close the context menu if required.
RegisterNetEvent('boii_ui:cl:close_action_menu', function()
    SetNuiFocus(false, false)
    active = false
    SendNUIMessage({ action = 'close_action_menu' })
end)

--- @section Exports

--- Opens a menu.
exports('action_menu', open_menu)

--- @section Test stuff

local test_menu = {
    {
        label = 'Level 1',
        icon = 'fa-solid fa-sitemap',
        colour = 'red',
        submenu = {
            {
                label = 'Level 2',
                icon = 'fa-solid fa-operations',
                colour = 'orange',
                submenu = {
                    {
                        label = 'Level 3',
                        icon = 'fa-solid fa-truck-moving',
                        colour = 'yellow',
                        submenu = {
                            {
                                label = 'Level 4',
                                icon = 'fa-solid fa-boxes',
                                colour = 'green',
                                submenu = {
                                    {
                                        label = 'Level 5',
                                        icon = 'fa-solid fa-map-marked-alt',
                                        colour = 'blue',
                                        submenu = {
                                            {
                                                label = 'Confirm',
                                                icon = 'fa-solid fa-check-circle',
                                                colour = 'purple',
                                                action_type = 'client',
                                                action = 'example_event',
                                                params = { example_param = 'example_value' }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    {
        label = 'Quick Access',
        icon = 'fa-solid fa-toolbox',
        action_type = 'client',
        action = 'boii_target:cl:open_toolbox',
        params = {}
    }
}

RegisterCommand('testactionmenu', function()
    SetNuiFocus(true, true)
    exports.boii_ui:action_menu(test_menu)
end, false)

File: html/scripts/classes/actionmenu.js
/*
     ____   ____ _____ _____   _   _____  ________      ________ _      ____  _____  __  __ ______ _   _ _______ 
    |  _ \ / __ \_   _|_   _| | | |  __ \|  ____\ \    / /  ____| |    / __ \|  __ \|  \/  |  ____| \ | |__   __|
    | |_) | |  | || |   | |   | | | |  | | |__   \ \  / /| |__  | |   | |  | | |__) | \  / | |__  |  \| |  | |   
    |  _ <| |  | || |   | |   | | | |  | |  __|   \ \/ / |  __| | |   | |  | |  ___/| |\/| |  __| | . ` |  | |   
    | |_) | |__| || |_ _| |_  | | | |__| | |____   \  /  | |____| |___| |__| | |    | |  | | |____| |\  |  | |   
    |____/ \____/_____|_____| | | |_____/|______|   \/   |______|______\____/|_|    |_|  |_|______|_| \_|  |_|   
                              | |                                                                                
                              |_|               UI ELEMENTS
*/

class ActionMenu {
    constructor() {
        this.menu_stack = [];
        this.current_menu = [];
        this.is_menu_active = false;
        this.create_menu_container();
    }

    create_menu_container() {
        this.menu_container = $('<div>').addClass('action_menu_container');
        $('#main_container').append(this.menu_container);
    }

    create_menu(actions) {
        this.menu_stack.push(actions);
        this.current_menu = actions;
        this.render_menu(actions);
        this.is_menu_active = true;
    }

    render_menu(actions) {
        this.menu_container.empty();
        let menu_left = $('<div>').addClass('actions left');
        let menu_right = $('<div>').addClass('actions right');
        let center_button = $('<div>').addClass('center_button').html('<i class="fa fa-rotate-left"></i>');
        center_button.on('click', () => this.handle_center_button_click());

        actions.forEach((action, index) => {
            let action_html = $('<div>').addClass('action').data('id', action.label);
            let colour = action.colour || '#b4b4b4';
            let label_html = `<span class="label" style="border: 2px solid ${colour}; color: ${colour};">${action.label}</span>`;
            let icon_html = `<span class="icon" style="border: 2px solid ${colour}; color: ${colour};"><i class="${action.icon}"></i></span>`;
            action_html.on('click', () => this.handle_action_click(action));
            if (index % 2 === 0) {
                action_html.html(`<div class="label_container">${icon_html}${label_html}</div>`);
                menu_right.append(action_html);
            } else {
                action_html.html(`<div class="label_container">${label_html}${icon_html}</div>`);
                menu_left.append(action_html);
            }
        });
        this.adjust_action_positions(menu_left, menu_right);
        this.menu_container.append(menu_left, center_button, menu_right);
    }
    
    adjust_action_positions(menu_left, menu_right) {
        let count_left = menu_left.children().length;
        let count_right = menu_right.children().length;
        menu_left.children().first().css('transform', '');
        menu_left.children().last().css('transform', '');
        menu_right.children().first().css('transform', '');
        menu_right.children().last().css('transform', '');
        if (count_left > 2 || count_right > 2) {
            menu_left.children().first().css('transform', count_left > 2 ? 'translateX(10%)' : '');
            menu_left.children().last().css('transform', count_left > 2 ? 'translateX(10%)' : '');
            menu_right.children().first().css('transform', count_right > 2 ? 'translateX(-10%)' : '');
            menu_right.children().last().css('transform', count_right > 2 ? 'translateX(-10%)' : '');
        }
    }

    handle_action_click(action) {
        if (action.submenu) {
            this.create_menu(action.submenu);
        } else {
            $.post(`https://${GetParentResourceName()}/trigger_event`, JSON.stringify({
                action_type: action.action_type,
                action: action.action,
                params: action.params
            }));
            if (this.menu_stack.length === 1) {
                this.is_menu_active = false;
                this.close();
            }
        }
    }

    handle_center_button_click() {
        if (this.menu_stack.length > 1) {
            this.menu_stack.pop();
            this.render_menu(this.menu_stack[this.menu_stack.length - 1]);
        } else {
            this.close();
        }
    }

    close() {
        this.menu_container.empty();
        this.is_menu_active = false;
        $.post(`https://${GetParentResourceName()}/close_action_menu`, JSON.stringify({}));
    }
}

/*
$(document).ready(function() {
    const test_menu = [
        {
            label: "Level 1",
            icon: "fa-solid fa-sitemap",
            colour: 'red',
            submenu: [
                {
                    label: "Level 2",
                    icon: "fa-solid fa-operations",
                    colour: 'orange',
                    submenu: [
                        {
                            label: "Level 3",
                            icon: "fa-solid fa-truck-moving",
                            colour: 'yellow',
                            submenu: [
                                {
                                    label: "Level 4",
                                    icon: "fa-solid fa-boxes",
                                    colour: 'green',
                                    submenu: [
                                        {
                                            label: "Level 5",
                                            icon: "fa-solid fa-map-marked-alt",
                                            colour: 'blue',
                                            submenu: [
                                                {
                                                    label: "Confirm",
                                                    icon: "fa-solid fa-check-circle",
                                                    colour: 'purple',
                                                    action_type: "client_event",
                                                    action: 'boii_target:cl:confirm_delivery',
                                                    params: { delivery_id: 1234 }
                                                }
                                            ]
                                        }
                                    ]
                                }
                            ]
                        }
                    ]
                }
            ]
        },
        {
            label: "Quick Access",
            icon: "fa-solid fa-toolbox",
            action_type: "client_event",
            action: 'boii_target:cl:open_toolbox',
            params: {}
        }
    ];

    const test_action_menu = new ActionMenu();
    test_action_menu.create_menu(test_menu);
});
*/
